docling_core/types/experimental/document.py

```diff
@@ -0,0 +1,1106 @@
+                    end_row_offset_idx=i + 1,
+                    start_col_offset_idx=j,
+                    end_col_offset_idx=j + 1,
+                )
+                for j in range(self.num_cols)
+            ]
+            for i in range(self.num_rows)
+        ]
+
+        # Overwrite cells in table data for which there is actual cell content.
+        for cell in self.table_cells:
+            for i in range(
+                min(cell.start_row_offset_idx, self.num_rows),
+                min(cell.end_row_offset_idx, self.num_rows),
+            ):
+                for j in range(
+                    min(cell.start_col_offset_idx, self.num_cols),
+                    min(cell.end_col_offset_idx, self.num_cols),
+                ):
+                    table_data[i][j] = cell
+
+        return table_data
+
+
+class FileInfo(BaseModel):
+    """FileInfo."""
+
+    filename: str
+    document_hash: str
+
+
+class RefItem(BaseModel):
+    """RefItem."""
+
+    cref: str = Field(alias="$ref")
+
+    # This method makes RefItem compatible with DocItem
+    def get_ref(self):
+        """get_ref."""
+        return self
+
+    model_config = ConfigDict(
+        populate_by_name=True,
+    )
+
+    def resolve(self, doc: "DoclingDocument"):
+        """resolve."""
+        path_components = self.cref.split("/")
+        if len(path_components) > 2:
+            _, path, index_str = path_components
+        else:
+            _, path = path_components
+            index_str = None
+
+        if index_str:
+            index = int(index_str)
+            obj = doc.__getattribute__(path)[index]
+        else:
+            obj = doc.__getattribute__(path)
+
+        return obj
+
+
+class ImageRef(BaseModel):
+    """ImageRef."""
+
+    format: str  # png, etc.
+    dpi: int  # ...
+    size: Size
+    uri: AnyUrl
+
+
+class ProvenanceItem(BaseModel):
+    """ProvenanceItem."""
+
+    page_no: int
+    bbox: BoundingBox
+    charspan: Tuple[int, int]
+
+
+class NodeItem(BaseModel):
+    """NodeItem."""
+
+    dloc: str  # format spec ({document_hash}{json-path})
+    parent: Optional[RefItem] = None
+    children: List[RefItem] = []
+
+    def get_ref(self):
+        """get_ref."""
+        return RefItem(cref=f"#{self.dloc.split('#')[1]}")
+
+    @computed_field  # type: ignore
+    @property
+    def hash(self) -> Uint64:  # TODO align with hasher on deepsearch-glm
+        """hash."""
+        if not len(self.dloc):
+            return 0
+        hash_object = hashlib.sha256(self.dloc.encode("utf-8"))
+
+        # Convert the hash to an integer
+        hash_int = int.from_bytes(hash_object.digest(), "big")
+
+        # Mask it to fit within 64 bits
+        return Uint64(hash_int & 0xFFFFFFFFFFFFFFFF)  # 64-bit unsigned integer mask
+
+
+class GroupItem(NodeItem):  # Container type, can't be a leaf node
+    """GroupItem."""
+
+    name: Optional[str] = None
+    label: GroupLabel = GroupLabel.UNSPECIFIED
```

Comment:
In other (most) places we have explicit typing with subclasses etc, but here we have a single type with a label instead. Maybe we want to use something like a [discriminated union](https://docs.pydantic.dev/latest/concepts/unions/#discriminated-unions)?

Reply:
Can you provide an example of how that would apply here? I don't see it yet.